pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java

```diff
@@ -39,168 +47,204 @@
-    // will consume the EOS block and computeReady() will never remove the mailbox
-    // from the _seenMail set.
-    //
-    // TODO: fix the memory leak by adding a close(opChain) callback
-    _seenMail.add(mailbox);
-    trace("got mail for " + mailbox);
+  public void deregister(OpChain operatorChain) {
+    _lock.lock();
+    try {
+      _aliveChains.remove(operatorChain.getId());
+      // it could be that the onDataAvailable callback was called when the OpChain was executing, in which case there
+      // could be a dangling entry in _seenMail.
+      _seenMail.remove(operatorChain.getId());
+    } finally {
+      _lock.unlock();
+    }
   }
 
   @Override
-  public boolean hasNext() {
-    if (!_ready.isEmpty()) {
-      return true;
+  public void yield(OpChain operatorChain) {
+    long releaseTs = _ticker.get() + _releaseTimeoutMs;
+    _lock.lock();
+    try {
+      // It could be that this OpChain received data before it could be yielded completely. In that case, mark it ready
+      // to get it scheduled asap.
+      if (_seenMail.contains(operatorChain.getId())) {
+        _seenMail.remove(operatorChain.getId());
+        _ready.add(operatorChain);
+        return;
+      }
+      _available.put(operatorChain.getId(), releaseTs);
+    } finally {
+      _lock.unlock();
     }
-    computeReady();
-    return !_ready.isEmpty();
   }
 
   @Override
-  public OpChain next() {
-    OpChain op = _ready.poll();
-    trace("Polled " + op);
+  public void onDataAvailable(MailboxIdentifier mailbox) {
+    // TODO: Should we add an API in MailboxIdentifier to get the requestId?
+    OpChainId opChainId = new OpChainId(Long.parseLong(mailbox.getJobId().split("_")[0]),
+        mailbox.getReceiverStageId());
+    // If this chain isn't alive as per the scheduler, don't do anything. If the OpChain is registered after this, it
+    // will anyways be scheduled to run since new OpChains are run immediately.
+    if (!_aliveChains.containsKey(opChainId)) {
+      trace("got mail, but the OpChain is not registered so ignoring the event " + mailbox);
+      return;
+    }
+    _lock.lock();
+    try {
+      if (!_aliveChains.containsKey(opChainId)) {
+        return;
+      }
+      if (_available.containsKey(opChainId)) {
+        _available.remove(opChainId);
+        _ready.offer(_aliveChains.get(opChainId));
+      } else {
+        // There are two cases here:
+        // 1. OpChain is in the _ready queue: the next time it gets polled, we'll remove the _seenMail entry.
+        // 2. OpChain is running: the next time yield is called for it, we'll check against _seenMail and put it back
+        //    in the _ready queue again.
+        _seenMail.add(opChainId);
+      }
+    } finally {
+      _lock.unlock();
+    }
+    trace("got mail for " + mailbox);
+  }
+
+  @Override
+  public OpChain next(long time, TimeUnit timeUnit) throws InterruptedException {
+    OpChain op = _ready.poll(time, timeUnit);
```

Comment:
Shall we check if the OpChain is still registered in _aliveChains?

Reply:
We could add a Precondition to catch that state, however by design it is guaranteed by the RoundRobinScheduler because:

1. OpChain can only be added to the _ready queue after it is registered.
2. OpChain that is running will never be in the _ready or _available collections.
3. OpChain can only be de-registered if it is running.